### DISCUSSION TITLE
CP-SAT get's stuck when trying to maximize sum of boolean variables that are all added as assumptions

### DISCUSSION THREAD

**Discussion by Mapiarz** [google/or-tools#3630] [CP-SAT questions] [answered]
Hi again!

**What version of OR-Tools and what language are you using?**
Version: v9.5
Language: Python

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
CP-SAT

**What operating system (Linux, Windows, ...) and version?**
Ubuntu 18.04 LTS

**Problem**
Solver get's indefinitely stuck and does not come up with any solution when the objective is set to maximize the sum of boolean variables which all have been added as assumption to the model. Below is an example model (real world, not minimal example) and solving log. 

Model: https://www.dropbox.com/s/7s5o7ixk9izmsyw/model.proto?dl=1
Log: https://www.dropbox.com/s/rzb1nnbo2pn8hh9/solver.log?dl=1

I've let the solver run at most for an hour and a half and it made no difference. When at least one of the variables is not added as an assumption, the solver comes up with a solution as expected. I do understand that setting such an objective when all the variables are added as assumptions does not really tell us anything, but the solver getting stuck caused a great deal of confusion for me and it took me quite a while to figure out what's causing it. I would expect the solver to come up with a solution or erroring out telling that the objective is moot.

What are your thoughts?

**Comment by lperron** [comment]
Have you looked at the log: 

Warning: solving with assumptions was requested in a non-fully supported setting.
We will assumes these assumptions true while solving, but if the model is infeasible, you will not get a useful 'sufficient_assumptions_for_infeasibility' field in the response, it will include all assumptions.

Objective and assumptions are incompatible.

**Comment by lperron** [comment]
So, 
minimizing the sum of Booleans, or using assumptions are doing the same thing. The minimization is exact. The assumptions are heuristics.

If I remove assumptions, I can solve. With 16 workers, the problem is not to complex

```
CpSolverResponse summary:
status: OPTIMAL
objective: 1993
best_bound: 1993
booleans: 6370
conflicts: 128325
branches: 200888
propagations: 4333409
integer_propagations: 6812319
restarts: 12754
lp_iterations: 111
walltime: 113.055
usertime: 113.055
deterministic_time: 289
gap_integral: 2.41023
solution_fingerprint: 0x5779e1eae315ab58
```

**Comment by Mapiarz** [comment]
Thanks for your response @lperron, helpful as always.

First of all, I wasn't aware that having an objective in a model is incompatible with assumptions. Perhaps that should be protected against programmatically by raising an error?

I do not use or rely on `sufficient_assumptions_for_infeasibility` so this incompatibility was never an issue for me so far.

Effectively, I was using `AddAssumption(lit)` as a shortcut for `Add(lit == 1)`. I was easily able to take the assumptions out of the picture entirely and just do `Add(lit == 1)`. Doing that still leads to the solver getting stuck and not come up with any solution in the same fashion as initially:
```
Starting CP-SAT solver v9.5.2237
Parameters: max_time_in_seconds: 450 log_search_progress: true num_workers: 4

Initial optimization model '': (model_fingerprint: 0x17f955ab0e9f41c4)
#Variables: 7193 (#bools:1995 in objective)
  - 6989 Booleans in [0,1]
  - 34 in [0,5]
  - 170 in [0,7]
#kLinMax: 304 (#expressions: 608)
#kLinear0: 51 (#enforced: 51)
#kLinear1: 2857 (#enforced: 862 #multi: 134)
#kLinear2: 99 (#enforced: 99)
#kLinear3: 19 (#enforced: 19)
#kLinearN: 1849 (#enforced: 1511) (#terms: 38561)

Starting presolve at 0.00s
[ExtractEncodingFromLinear] #potential_supersets=720 #potential_subsets=107 #at_most_one_encodings=0 #exactly_one_encodings=0 #unique_terms=0 #multiple_terms=0 #literals=0 time=0.000188776s
[Probing] deterministic_time: 0.0070643 (limit: 1) wall_time: 0.016144 (1943/1943)
[Probing]  - new fixed Boolean: 4 (5104/7043)
[Probing]  - new integer bounds: 22
[Probing]  - new binary clause: 1456
[Probing] implications and bool_or (work_done=24213).
[SAT presolve] num removable Booleans: 5386 / 7193
[SAT presolve] num trivial clauses: 0
[SAT presolve] [0s] clauses:47 literals:113 vars:105 one_side_vars:105 simple_definition:0 singleton_clauses:0
[SAT presolve] [3.8432e-05s] clauses:47 literals:113 vars:105 one_side_vars:105 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.000464373s] clauses:47 literals:113 vars:105 one_side_vars:105 simple_definition:0 singleton_clauses:0
[MaxClique] Merged 633(4934 literals) into 316(3812 literals) at_most_ones.
[DetectDuplicateConstraints] #duplicates=4 #without_enforcements=0 time=0.000629348s
[DetectDominatedLinearConstraints] #relevant_constraints=205 #work_done=29843 #num_inclusions=89 #num_redundant=24 time=0.00179034s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0 #nz_reduction=0 time=0.000323917s
[ProcessSetPPC] #relevant_constraints=598 #num_inclusions=370 work=48323 time=0.00207786s
[Probing] deterministic_time: 0.00666229 (limit: 1) wall_time: 0.0145125 (1938/1938)
[Probing]  - new binary clause: 1456
[Probing] implications and bool_or (work_done=19353).
[SAT presolve] num removable Booleans: 5388 / 7193
[SAT presolve] num trivial clauses: 0
[SAT presolve] [0s] clauses:22 literals:125 vars:122 one_side_vars:122 simple_definition:0 singleton_clauses:0
[SAT presolve] [3.3478e-05s] clauses:22 literals:125 vars:122 one_side_vars:122 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.000457258s] clauses:22 literals:125 vars:122 one_side_vars:122 simple_definition:0 singleton_clauses:0
[DetectDuplicateConstraints] #duplicates=0 #without_enforcements=0 time=0.000623285s
[DetectDominatedLinearConstraints] #relevant_constraints=152 #work_done=15828 #num_inclusions=18 #num_redundant=0 time=0.00104467s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0 #nz_reduction=0 time=0.000314325s
[ProcessSetPPC] #relevant_constraints=428 #num_inclusions=189 work=34305 time=0.00161553s

Presolve summary:
  - 131 affine relations were detected.
  - rule 'TODO dual: only one blocking constraint?' was applied 1723 times.
  - rule 'TODO dual: only one unspecified blocking constraint?' was applied 4040 times.
  - rule 'TODO linear inclusion: superset is equality' was applied 3 times.
  - rule 'TODO linear2: contains a Boolean.' was applied 306 times.
  - rule 'affine: new relation' was applied 131 times.
  - rule 'at_most_one: empty or all false' was applied 277 times.
  - rule 'at_most_one: removed literals' was applied 744 times.
  - rule 'at_most_one: size one' was applied 88 times.
  - rule 'at_most_one: transformed into max clique.' was applied 1 time.
  - rule 'at_most_one: x and not(x)' was applied 6 times.
  - rule 'bool_and: non-reified.' was applied 14 times.
  - rule 'bool_and: x => x' was applied 121 times.
  - rule 'bool_or: implications' was applied 154 times.
  - rule 'deductions: 134 stored' was applied 1 time.
  - rule 'dual: fix variable' was applied 1 time.
  - rule 'duplicate: merged rhs of linear constraint' was applied 1 time.
  - rule 'duplicate: removed constraint' was applied 4 times.
  - rule 'enforcement: false literal' was applied 109 times.
  - rule 'enforcement: true literal' was applied 2505 times.
  - rule 'exactly_one: removed literals' was applied 367 times.
  - rule 'exactly_one: singleton' was applied 121 times.
  - rule 'exactly_one: size two' was applied 2 times.
  - rule 'exactly_one: x and not(x)' was applied 13 times.
  - rule 'lin_max: converted to equality' was applied 197 times.
  - rule 'lin_max: removed exprs' was applied 197 times.
  - rule 'lin_max: target domain reduced' was applied 61 times.
  - rule 'linear + amo: extracted enforcement literal' was applied 96 times.
  - rule 'linear + amo: simplified enforcement list' was applied 1 time.
  - rule 'linear + amo: trivial linear constraint' was applied 21 times.
  - rule 'linear inclusion: redundant containing constraint' was applied 1 time.
  - rule 'linear inclusion: redundant included constraint' was applied 20 times.
  - rule 'linear inclusion: subset + singleton is equality' was applied 3 times.
  - rule 'linear1: is boolean implication' was applied 134 times.
  - rule 'linear1: without enforcement' was applied 33 times.
  - rule 'linear: always true' was applied 420 times.
  - rule 'linear: divide by GCD' was applied 33 times.
  - rule 'linear: empty' was applied 2980 times.
  - rule 'linear: fixed or dup variables' was applied 4693 times.
  - rule 'linear: infeasible' was applied 3 times.
  - rule 'linear: negative clause' was applied 163 times.
  - rule 'linear: positive at most one' was applied 843 times.
  - rule 'linear: positive clause' was applied 1 time.
  - rule 'linear: positive equal one' was applied 245 times.
  - rule 'linear: reduced variable domains' was applied 3072 times.
  - rule 'linear: reduced variable domains in derived constraint' was applied 80 times.
  - rule 'linear: remapped using affine relations' was applied 13 times.
  - rule 'linear: simplified rhs' was applied 1071 times.
  - rule 'linear: simplify using approximate gcd' was applied 3 times.
  - rule 'linear: singleton column' was applied 39 times.
  - rule 'linear: variable substitution 0' was applied 2 times.
  - rule 'objective: variable not used elsewhere' was applied 1995 times.
  - rule 'presolve: 5107 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 2 times.
  - rule 'probing: bool_or reduced to implication' was applied 1 time.
  - rule 'setppc: bool_or in at_most_one.' was applied 120 times.
  - rule 'setppc: removed dominated constraints' was applied 44 times.

Presolved optimization model '': (model_fingerprint: 0x3416927186b0ddb1)
#Variables: 1797 ( in objective)
  - 1690 Booleans in [0,1]
  - 18 in [0,2]
  - 26 in [0,3]
  - 32 in [0,4]
  - 6 in [0,5]
  - 21 in [0,6]
  - 4 in [0,7]
#kAtMostOne: 144 (#literals: 1665)
#kBoolAnd: 9 (#enforced: 9) (#literals: 9)
#kBoolOr: 13 (#literals: 107)
#kExactlyOne: 119 (#literals: 1563)
#kLinMax: 107 (#expressions: 214)
#kLinear3: 5
#kLinearN: 152 (#enforced: 5 #multi: 2) (#terms: 3823)

Preloading model.
#Bound   0.13s best:-inf  next:[1995,1995] initial_domain
#Model   0.13s var:1797/1797 constraints:549/549

Starting  search at 0.13s with 4 workers.
2 full subsolvers: [default_lp, no_lp]
Interleaved subsolvers: [feasibility_pump, rnd_var_lns_default, rnd_cst_lns_default, graph_var_lns_default, graph_cst_lns_default, rins_lns_default, rens_lns_default]
```

To recap: I have a number of boolean variables, for each of them I do `Add(variable == 1)` and then I set the objective to be `Maximize(sum(variables))`.

**Comment by lperron** [comment]
use more workers. Try 8.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by Mapiarz** [comment]
@lperron I have tried using 8 workers, it does indeed help. The solver is able to find solution (or prove infeasibility) almost instantly. Actually, any number of workers greater than 4 results in the solution being found almost instantly. Surely this cannot be explained by a simple performance gain, can it? This is pretty worrisome.

**Comment by lperron** [comment, marked as answer]
The solver is designed to work with 8+ workers, unless for trivial models.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
